consensus/src/processes/pruning_proof/mod.rs

```diff
@@ -614,45 +714,285 @@ impl PruningProofManager {
+                topological_heap.push(Reverse(SortableBlock {
+                    hash: child,
+                    // It's important to use here blue work and not score so we can iterate the heap in a way that respects the topology
+                    blue_work: self.headers_store.get_header(child).unwrap().blue_work, // TODO: Maybe add to compact data?
+                }));
+            }
+
+            let mut has_required_block = root == required_block;
+            loop {
+                let Some(current) = topological_heap.pop() else {
+                    break;
+                };
+                let current_hash = current.0.hash;
+                if !visited.insert(current_hash) {
+                    continue;
+                }
+
+                if !self.reachability_service.is_dag_ancestor_of(current_hash, pp) {
+                    // We don't care about blocks in the antipast of the pruning point
+                    continue;
+                }
+
+                if !has_required_block && current_hash == required_block {
+                    has_required_block = true;
+                }
+
+                let relevant_parents: Box<[Hash]> = relations_service
+                    .get_parents(current_hash)
+                    .unwrap()
+                    .iter()
+                    .copied()
+                    .filter(|parent| self.reachability_service.is_dag_ancestor_of(root, *parent))
+                    .collect();
+                let current_gd = gd_manager.ghostdag(&relevant_parents);
+                ghostdag_store.insert(current_hash, Arc::new(current_gd)).unwrap();
+                for child in relations_service.get_children(current_hash).unwrap().read().iter().copied() {
+                    topological_heap.push(Reverse(SortableBlock {
+                        hash: child,
+                        // It's important to use here blue work and not score so we can iterate the heap in a way that respects the topology
+                        blue_work: self.headers_store.get_header(child).unwrap().blue_work, // TODO: Maybe add to compact data?
+                    }));
+                }
+            }
+
+            if has_required_block
+                && (root == self.genesis_hash || ghostdag_store.get_blue_score(selected_tip).unwrap() >= required_level_depth)
+            {
+                break Ok((ghostdag_store, selected_tip, root));
+            }
+
+            tries += 1;
+            if finished_headers {
+                panic!("Failed to find sufficient root for level {level} after {tries} tries. Headers below the current depth of {required_level_0_depth} are already pruned")
+            }
+            required_level_0_depth <<= 1;
+            warn!("Failed to find sufficient root for level {level} after {tries} tries. Retrying again to find with depth {required_level_0_depth}");
+        }
+    }
+
+    fn calc_gd_for_all_levels(
```

Comment:
Need some comments here. From what I can understand, this fills a temp DB with the ghostdag data. I assume that DB is used for pruning proof only, and that this DB is shortlived.